service/matching/user_data_manager.go

```diff
@@ -0,0 +1,431 @@
+}
+
+// Stop does not unload the partition from matching engine. It is intended to be called by matching engine when
+// unloading the partition. For stopping and unloading a partition call unloadFromEngine instead.
+func (m *userDataManager) Stop() {
+	m.goroGroup.Cancel()
+	// Set user data state on stop to wake up anyone blocked on the user data changed channel.
+	m.setUserDataState(userDataClosed, nil)
+}
+
+// GetUserData returns the versioning data for this task queue. Do not mutate the returned pointer, as doing so
+// will cause cache inconsistency.
+func (m *userDataManager) GetUserData() (*persistencespb.VersionedTaskQueueUserData, chan struct{}, error) {
+	m.Lock()
+	defer m.Unlock()
+	return m.getUserDataLocked()
+}
+
+func (m *userDataManager) getUserDataLocked() (*persistencespb.VersionedTaskQueueUserData, chan struct{}, error) {
+	switch m.userDataState {
+	case userDataEnabled:
+		return m.userData, m.userDataChanged, nil
+	case userDataClosed:
+		return nil, nil, errTaskQueueClosed
+	default:
+		// shouldn't happen
+		return nil, nil, serviceerror.NewInternal("unexpected user data enabled state")
+	}
+}
+
+func (m *userDataManager) setUserDataLocked(userData *persistencespb.VersionedTaskQueueUserData) {
+	m.userData = userData
+	close(m.userDataChanged)
+	m.userDataChanged = make(chan struct{})
+}
+
+// Sets user data enabled/disabled and marks the future ready (if it's not ready yet).
+// userDataState controls whether GetUserData return an error, and which.
+// futureError is the error to set on the ready future. If this is non-nil, the task queue will
+// be unloaded.
+// Note that this must only be called from a single goroutine since the Ready/Set sequence is
+// potentially racy otherwise.
+func (m *userDataManager) setUserDataState(userDataState userDataState, futureError error) {
+	// Always set state enabled/disabled even if we're not setting the future since we only set
+	// the future once but the enabled/disabled state may change over time.
+	m.Lock()
+	defer m.Unlock()
+
+	if userDataState != m.userDataState && m.userDataState != userDataClosed {
+		m.userDataState = userDataState
+		close(m.userDataChanged)
+		m.userDataChanged = make(chan struct{})
+	}
+
+	if !m.userDataReady.Ready() {
+		m.userDataReady.Set(struct{}{}, futureError)
+	}
+}
+
+func (m *userDataManager) loadUserData(ctx context.Context) error {
+	ctx = m.callerInfoContext(ctx)
+	err := m.loadUserDataFromDB(ctx)
+	m.setUserDataState(userDataEnabled, err)
+	return nil
+}
+
+func (m *userDataManager) userDataFetchSource() (*tqid.NormalPartition, error) {
+	switch p := m.partition.(type) {
+	case *tqid.NormalPartition:
+		degree := m.config.ForwarderMaxChildrenPerNode()
+		parent, err := p.ParentPartition(degree)
+		if err == tqid.ErrNoParent { // nolint:goerr113
+			// we're the root activity task queue, ask the root workflow task queue
+			return p, nil
+		} else if err != nil {
+			// invalid degree
+			return nil, err
+		}
+		return parent, nil
```

Comment:
it would be cleaner to replace the type field here, and then the caller should use the type from the return value (instead of assuming workflow)